clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java

```diff
@@ -81,11 +119,23 @@ public void initialize(URI name, Configuration conf) throws IOException {
      * regardless of the given file path.
      */
     @Override
-    public FSDataInputStream open(Path path, int i) throws IOException {
-        LOG.debug("$$$$$$$$$$$$$$$$$$$$$$$$$$$$ Calling open method for: {} $$$$$$$$$$$$$$$$$$$$$$$$$$$$", path.getName());
-        String strToWrite = "abc";
-        MyInputStream inputStream = new MyInputStream(strToWrite);
-        return new FSDataInputStream(inputStream);
+    public FSDataInputStream open(Path path, int bufSize) throws IOException {
+	try {
+	    LOG.debug("$$$$$$$$$$$$$$$$$$$$$$$$$$$$ open(" + path.getName() + ") $$$$$$$$$$$$$$$$$$$$$$$$$$$$");
+
+	    ObjectsApi objects = new ObjectsApi(apiClient);
+	    ParsedPath pp = parsePath(path);
+	    ObjectStats stats = objects.statObject(pp.repo, pp.ref, pp.path);
+	    URI physicalUri = translateUri(new URI(stats.getPhysicalAddress()));
+
+	    Path physicalPath = new Path(physicalUri.toString());
+	    FileSystem physicalFs = physicalPath.getFileSystem(conf);
```

Comment:
This is the part where I thought we need to initialize the underlying filesystem once in the `initialize` and use it here and just call `open`.

Reply:
Yeah, I was hoping to do that too.  Disappointed that this is actually surprisingly difficult to do correctly.  Now according to [`S3AFileSystem.initialize`](https://github.com/grahamar/hadoop-aws/blob/c127dfbd70e6d166d6e6517680cf46c5e15738e9/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java#L167) (please look at the code), the repository is guaranteed to exist on the URL.  (TBH, I don't understand _why_, but we might assume as much).  So we _could_ generate the underlying file system at that point.

Let's circle back to this point once we have a working FS?  Because what this would do is (only) speed up multiple calls to `open` (and `create`, after #1858 ) when they occur on the same FileSystem.  I.e. it speeds up the already-fast part of multi-part file read/write operations -- if it works.

Opened #1906.